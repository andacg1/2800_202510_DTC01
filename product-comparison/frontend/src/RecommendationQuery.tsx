import React, { useEffect, useState } from "react";
import type { Product } from "./product.ts";

type RecommendationQueryProps = {
  className?: string;
  children?: React.ReactNode;
  products?: Product[];
};

type Recommendation = {
  recommendedProductId: string;
  recommendedProductTitle: string;
  reason: string;
};

type RecommendationResponse = {
  outputJson: Recommendation;
};

const RecommendationQuery = ({
  className,
  children,
  products,
}: RecommendationQueryProps) => {
  const [response, setResponse] = useState<Recommendation>();
  const [isLoading, setIsLoading] = useState(false);
  useEffect(() => {}, []);
  // I'm a retired snowboarded who just lost everything after my 7th divorce. What kind of gear should I buy?

  return (
    <div>
      <form
        style={{
          display: "flex",
          flexDirection: "column",
          alignItems: "end",
        }}
        onSubmit={async (e) => {
          e.preventDefault();
          const data = new FormData(e.currentTarget);
          const query = data.get("query");
          // TODO: define this port in the shopify CLI config, if possible
          const url = `${process.env.APP_BACKEND_URL}/recommend`;
          console.log(url);
          if (!url) {
            return console.error("Could not find APP_BACKEND_URL");
          }
          setIsLoading(true);
          const resp = await fetch(url, {
            method: "POST",
            body: JSON.stringify({
              query,
              products,
            }),
          });
          const body: RecommendationResponse = await resp.json();
          setIsLoading(false);
          setResponse(body.outputJson);
        }}
      >
        <textarea
          className="w-full"
          placeholder={"Tell us about your use case"}
          name="query"
          style={{ width: "100%" }}
        />
        <button type="submit" style={{ width: "100px", marginTop: "8px" }}>
          Submit
        </button>
      </form>
      {isLoading ? "Loading..." : null}
      {response?.reason ? <p>{response.reason}</p> : null}
    </div>
  );
};

export default RecommendationQuery;
