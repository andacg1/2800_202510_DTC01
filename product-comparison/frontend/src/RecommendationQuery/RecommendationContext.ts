import type React from "react";
import { createContext } from "react";
import type { Product } from "../product.ts";

/**
 * Props for the RecommendationQuery component
 * @typedef {Object} RecommendationQueryProps
 * @property {string} [className] - Optional CSS class name for styling
 * @property {React.ReactNode} [children] - Optional child elements
 * @property {Product[]} [products] - Array of products to generate recommendations from
 */
export type RecommendationQueryProps = {
  className?: string;
  children?: React.ReactNode;
  products?: Product[];
};

/**
 * Structure of a product recommendation
 * @typedef {Object} Recommendation
 * @property {string} recommendedProductId - ID of the recommended product
 * @property {string} recommendedProductTitle - Title of the recommended product
 * @property {string} reason - Explanation for why this product was recommended
 */
export type Recommendation = {
  recommendedProductId: string;
  recommendedProductTitle: string;
  reason: string;
};

/**
 * Response structure from the recommendation API
 * @typedef {Object} RecommendationResponse
 * @property {Recommendation} outputJson - The recommendation data
 */
export type RecommendationResponse = {
  outputJson: Recommendation;
};

/**
 * Type definition for the Recommendation context value
 * @typedef {Object} RecommendationContextType
 * @property {Recommendation | undefined} recommendation - Current recommendation or undefined
 * @property {React.Dispatch<React.SetStateAction<Recommendation | undefined>>} setRecommendation - Function to update recommendation
 * @property {string | undefined} query - Current search query
 * @property {React.Dispatch<React.SetStateAction<string | undefined>>} setQuery - Function to update search query
 */
export type RecommendationContextType = {
  recommendation: Recommendation | undefined;
  setRecommendation: React.Dispatch<
    React.SetStateAction<Recommendation | undefined>
  >;
  query: string | undefined;
  setQuery: React.Dispatch<React.SetStateAction<string | undefined>>;
};

/**
 * Default values for the Recommendation context
 */
export const defaultRecommendationContext: RecommendationContextType = {
  recommendation: {
    recommendedProductId: "",
    recommendedProductTitle: "",
    reason: "",
  },
  setRecommendation: () => {},
  query: "",
  setQuery: () => {},
};

/**
 * React context for managing product recommendations throughout the application
 * Provides access to recommendation state and query functionality
 */
export const RecommendationContext = createContext(
  defaultRecommendationContext,
);
