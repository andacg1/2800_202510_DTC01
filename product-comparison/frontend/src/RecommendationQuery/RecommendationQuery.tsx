import React, { useContext, useEffect, useState } from "react";
import {
  defaultRecommendationContext,
  Recommendation,
  RecommendationContext,
  RecommendationQueryProps,
  RecommendationResponse,
} from "./RecommendationContext.ts";

async function mockOpenApiResponse() {
  return {
    success: true,
    outputJson: {
      recommendedProductId: "9962241655059",
      recommendedProductTitle: "Copper Light",
      reason:
        "As a level 5 kobold working in dim mines, your primary issue is inadequate illumination from your head-candle. Of all the products you listed, the 'Copper Light' is the only item that is actually designed to produce light. Snowboards, wax, sofas, and plant pots do not provide any illumination for your underground activities. The 'Copper Light' is a bedside lamp, but it’s portable and much brighter than a typical kobold candle—making it the most logical upgrade for lighting up mine tunnels. Just make sure you find a way to power it underground!",
    },
    message: {
      id: "resp_681fca9d95d4819183c0eb7d807722830abf5240ac5e06eb",
      object: "response",
      created_at: 1746913949,
      status: "completed",
      error: null,
      incomplete_details: null,
      instructions: null,
      max_output_tokens: null,
      model: "gpt-4.1-2025-04-14",
      output: [
        {
          id: "msg_681fca9f324c8191af5b99dee4988ebe0abf5240ac5e06eb",
          type: "message",
          status: "completed",
          content: [
            {
              type: "output_text",
              annotations: [],
              text: '{"recommendedProductId":"9962241655059","recommendedProductTitle":"Copper Light","reason":"As a level 5 kobold working in dim mines, your primary issue is inadequate illumination from your head-candle. Of all the products you listed, the \'Copper Light\' is the only item that is actually designed to produce light. Snowboards, wax, sofas, and plant pots do not provide any illumination for your underground activities. The \'Copper Light\' is a bedside lamp, but it’s portable and much brighter than a typical kobold candle—making it the most logical upgrade for lighting up mine tunnels. Just make sure you find a way to power it underground!"}',
            },
          ],
          role: "assistant",
        },
      ],
      parallel_tool_calls: true,
      previous_response_id: null,
      reasoning: {
        effort: null,
        summary: null,
      },
      service_tier: "default",
      store: true,
      temperature: 1,
      text: {
        format: {
          type: "json_schema",
          description: null,
          name: "recommendation",
          schema: {
            type: "object",
            properties: {
              recommendedProductId: { type: "string" },
              recommendedProductTitle: { type: "string" },
              reason: { type: "string" },
            },
            required: [
              "recommendedProductId",
              "recommendedProductTitle",
              "reason",
            ],
            additionalProperties: false,
          },
          strict: true,
        },
      },
      tool_choice: "auto",
      tools: [],
      top_p: 1,
      truncation: "disabled",
      usage: {
        input_tokens: 21795,
        input_tokens_details: { cached_tokens: 0 },
        output_tokens: 138,
        output_tokens_details: { reasoning_tokens: 0 },
        total_tokens: 21933,
      },
      user: null,
      metadata: {},
      output_text:
        '{"recommendedProductId":"9962241655059","recommendedProductTitle":"Copper Light","reason":"As a level 5 kobold working in dim mines, your primary issue is inadequate illumination from your head-candle. Of all the products you listed, the \'Copper Light\' is the only item that is actually designed to produce light. Snowboards, wax, sofas, and plant pots do not provide any illumination for your underground activities. The \'Copper Light\' is a bedside lamp, but it’s portable and much brighter than a typical kobold candle—making it the most logical upgrade for lighting up mine tunnels. Just make sure you find a way to power it underground!"}',
    },
  };
}

const RecommendationQuery = ({
  className,
  children,
  products,
}: RecommendationQueryProps) => {
  const { recommendation, setRecommendation, query, setQuery } = useContext(
    RecommendationContext,
  );
  const isLoading = query && (!recommendation || !recommendation?.reason);

  return (
    <div className="mb-12">
      <form
        style={{
          display: "flex",
          flexDirection: "column",
          alignItems: "end",
        }}
        className="flex flex-col items-end mb-6"
        onSubmit={async (e) => {
          e.preventDefault();
          const queryInput = e.currentTarget.elements.namedItem(
            "query",
          ) as HTMLTextAreaElement;
          console.log(e.currentTarget.elements);
          const data = new FormData(e.currentTarget);
          const queryData = data.get("query");
          // TODO: define this port in the shopify CLI config, if possible
          if (!(typeof queryData === "string")) {
            return console.error("Query cannot be a File");
          }
          if (queryInput && queryInput?.value) {
            queryInput.value = "";
          }
          setQuery(queryData || "");
          setRecommendation(defaultRecommendationContext.recommendation);
          if (!process.env.APP_BACKEND_URL) {
            return console.error("Could not find APP_BACKEND_URL");
          }
          const url = `${process.env.APP_BACKEND_URL}/recommend`;
          console.log(url);

          let body: RecommendationResponse;

          const MOCK_OPENAI_REQUEST = false;
          if (MOCK_OPENAI_REQUEST) {
            body = await mockOpenApiResponse();
          } else {
            const resp = await fetch(url, {
              method: "POST",
              body: JSON.stringify({
                query: queryData,
                products,
              }),
            });
            body = await resp.json();
          }
          setRecommendation(body.outputJson);
        }}
      >
        <textarea
          className="textarea w-full text-2xl placeholder:opacity-40"
          placeholder={"Tell us about your use case"}
          name="query"
        />
        <button
          type="submit"
          className={"btn btn-primary"}
          style={{ width: "100px", marginTop: "8px" }}
        >
          Submit
        </button>
      </form>
      {query ? (
        <div className="chat chat-end">
          <div className="chat-bubble chat-bubble-info max-w-prose">
            {query}
          </div>
        </div>
      ) : null}
      {isLoading ? (
        <span className="loading loading-dots loading-md"></span>
      ) : null}
      {recommendation?.reason ? (
        <div className="chat chat-start mb-1">
          <div className="chat-bubble max-w-prose">
            {recommendation?.reason}
          </div>
        </div>
      ) : null}
    </div>
  );
};

export default RecommendationQuery;
