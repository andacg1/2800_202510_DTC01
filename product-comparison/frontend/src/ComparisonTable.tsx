import { faCheck, faX } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import React, { useContext, useEffect, useState } from "react";
import {
  getMockLocation,
  isAvailable,
  type LocationData,
  type Product,
} from "./product.ts";
import { RecommendationContext } from "./RecommendationQuery/RecommendationContext.ts";

type ComparisonTableProps = {
  className?: string;
  children?: React.ReactNode;
  products: Product[];
};

const ComparisonTable = ({
  className,
  children,
  products,
}: ComparisonTableProps) => {
  const [selectedProducts, setSelectedProducts] = useState<number[]>([]);

  const [userLocation, setUserLocation] = useState<LocationData>();

  const { recommendation, setRecommendation } = useContext(
    RecommendationContext,
  );

  const toggleProductSelection = (productId: number) => {
    setSelectedProducts((prev) => {
      if (prev.includes(productId)) {
        return prev.filter((id) => id !== productId);
      } else {
        // Limit to comparing 2 products maximum
        if (prev.length < 2) {
          return [...prev, productId];
        }
        return [prev[1], productId]; // Replace oldest selection
      }
    });
  };

  const ensureProductSelection = (productId: number) => {
    setSelectedProducts((prev) => {
      if (prev.includes(productId)) {
        return [...prev];
      } else {
        // Limit to comparing 2 products maximum
        if (prev.length < 2) {
          return [...prev, productId];
        }
        return [prev[1], productId]; // Replace oldest selection
      }
    });
  };

  // Get all unique spec keys across selected products
  const getAllSpecKeys = () => {
    const selectedProductData = products.filter((p) =>
      selectedProducts.includes(p.id),
    );
    const allKeys = new Set<string>();

    selectedProductData.forEach((product) => {
      Object.keys(product.specs).forEach((key) => {
        allKeys.add(key);
      });
    });

    return Array.from(allKeys);
  };

  useEffect(() => {
    (async () => {
      //const location = await getLocation();
      const location = await getMockLocation();
      setUserLocation(location);
    })();
  }, []);

  useEffect(() => {
    if (!recommendation?.recommendedProductId) {
      return;
    }
    const shortId = recommendation.recommendedProductId.slice(22);
    setSelectedProducts([Number(shortId)]);
    // ensureProductSelection(Number(shortId));
  }, [recommendation]);

  // Track comparison when two products are selected
  useEffect(() => {
    const trackComparison = async () => {
      // This would normally send data to your Shopify app backend
      if (selectedProducts.length === 2) {
        try {
          console.log(
            `Tracking comparison between ${selectedProducts[0]} and ${selectedProducts[1]}`,
          );
          // In a real implementation, you would call your API endpoint
          // await fetch('/api.product.comparison.track', {
          //   method: 'POST',
          //   headers: { 'Content-Type': 'application/json' },
          //   body: JSON.stringify({
          //     productAId: selectedProducts[0],
          //     productBId: selectedProducts[1],
          //     userId: 'anonymous',
          //     shop: window.location.hostname
          //   })
          // });
        } catch (err) {
          console.error("Failed to track comparison", err);
        }
      }
    };

    if (selectedProducts.length === 2) {
      trackComparison();
    }
  }, [selectedProducts]);

  return (
    <div className={className}>
      <div className="product-selection">
        <h3>Select Products to Compare (max 2)</h3>
        <div className="products-grid">
          {products.map((product) => (
            <div
              key={product.id}
              className={`product-card ${selectedProducts.includes(product.id) ? "selected" : ""}`}
              onClick={() => toggleProductSelection(product.id)}
            >
              <h4>{product.title}</h4>
              <div className="select-indicator">
                {selectedProducts.includes(product.id) ? "âœ“" : "+"}
              </div>
            </div>
          ))}
        </div>
      </div>
      {selectedProducts.length > 0 && (
        <div className="comparison-table">
          <h3>Comparison</h3>
          <table>
            <thead>
              <tr>
                <th>Specification</th>
                {selectedProducts.map((productId) => {
                  const product = products.find((p) => p.id === productId);
                  return <th key={productId}>{product?.title}</th>;
                })}
              </tr>
            </thead>
            <tbody>
              {getAllSpecKeys().map((specKey) => (
                <tr key={specKey}>
                  {specKey === "available_regions" && userLocation ? (
                    <td>{`Available in ${userLocation?.country_name || "N/A"}?`}</td>
                  ) : (
                    <td>{specKey.replace("_", " ")}</td>
                  )}
                  {selectedProducts.map((productId) => {
                    const product = products.find((p) => p.id === productId);
                    const specValue = product?.specs[specKey];
                    const isAvailableField = specKey === "available_regions";
                    if (
                      isAvailableField &&
                      userLocation &&
                      Array.isArray(specValue)
                    ) {
                      return (
                        <td
                          key={`${productId}-${specKey}`}
                          className="text-center"
                        >
                          {isAvailable(userLocation, specValue) ? (
                            <FontAwesomeIcon
                              icon={faCheck}
                              color={"rgba(31,255,0,0.5)"}
                              size={"2x"}
                              fixedWidth
                            />
                          ) : (
                            <FontAwesomeIcon
                              icon={faX}
                              color={"#b02525"}
                              size={"2x"}
                              fixedWidth
                            />
                          )}
                        </td>
                      );
                    }
                    return (
                      <td
                        key={`${productId}-${specKey}`}
                        className="text-center"
                      >
                        {Array.isArray(specValue)
                          ? specValue.join(", ")
                          : specValue?.toString() || "N/A"}
                      </td>
                    );
                  })}
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      )}
    </div>
  );
};

export default ComparisonTable;
