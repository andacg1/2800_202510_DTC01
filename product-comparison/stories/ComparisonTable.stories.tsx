import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import ComparisonTable from "../frontend/src/ComparisonTable";
import { RecommendationContext } from "../frontend/src/RecommendationQuery/RecommendationContext";

// Sample product data for the story
const sampleProducts = [
  {
    id: 1,
    title: "Product A",
    handle: "product-a",
    description: "Description for Product A",
    published_at: "2023-01-01",
    created_at: "2023-01-01",
    updated_at: "2023-01-01",
    vendor: "Vendor A",
    product_type: "Type A",
    tags: ["tag1", "tag2"],
    variants: [],
    images: [],
    options: [],
    shop: "shop.myshopify.com",
    status: "active",
    price: 100,
    compare_at_price: 120,
    inventory_quantity: 10,
    inventory_management: "shopify",
    inventory_policy: "deny",
    fulfillment_service: "manual",
    requires_shipping: true,
    taxable: true,
    weight: 1,
    weight_unit: "kg",
    specs: {
      color: "Red",
      size: "Medium",
      weight: "2kg",
      available_regions: ["US", "CA", "EU"],
    },
  },
  {
    id: 2,
    title: "Product B",
    handle: "product-b",
    description: "Description for Product B",
    published_at: "2023-01-01",
    created_at: "2023-01-01",
    updated_at: "2023-01-01",
    vendor: "Vendor B",
    product_type: "Type B",
    tags: ["tag1", "tag3"],
    variants: [],
    images: [],
    options: [],
    shop: "shop.myshopify.com",
    status: "active",
    price: 150,
    compare_at_price: 180,
    inventory_quantity: 5,
    inventory_management: "shopify",
    inventory_policy: "deny",
    fulfillment_service: "manual",
    requires_shipping: true,
    taxable: true,
    weight: 2,
    weight_unit: "kg",
    specs: {
      color: "Blue",
      size: "Large",
      weight: "3kg",
      available_regions: ["US", "EU"],
    },
  },
];

const meta = {
  title: "Product Comparison/ComparisonTable",
  component: ComparisonTable,
  parameters: {
    layout: "fullscreen",
  },
  decorators: [
    (Story) => (
      <div data-theme="light" style={{ padding: "1rem" }}>
        <RecommendationContext.Provider
          value={{
            recommendation: undefined,
            setRecommendation: () => {},
            query: undefined,
            setQuery: () => {},
          }}
        >
          <Story />
        </RecommendationContext.Provider>
      </div>
    ),
  ],
  tags: ["autodocs"],
} satisfies Meta<typeof ComparisonTable>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Empty: Story = {
  args: {
    products: [],
  },
};

export const WithTwoProducts: Story = {
  args: {
    products: sampleProducts,
  },
};

export const WithThreeProducts: Story = {
  args: {
    products: [
      ...sampleProducts,
      {
        id: 3,
        title: "Product C",
        handle: "product-c",
        description: "Description for Product C",
        published_at: "2023-01-01",
        created_at: "2023-01-01",
        updated_at: "2023-01-01",
        vendor: "Vendor C",
        product_type: "Type C",
        tags: ["tag1", "tag4"],
        variants: [],
        images: [],
        options: [],
        shop: "shop.myshopify.com",
        status: "active",
        price: 80,
        compare_at_price: 100,
        inventory_quantity: 15,
        inventory_management: "shopify",
        inventory_policy: "deny",
        fulfillment_service: "manual",
        requires_shipping: true,
        taxable: true,
        weight: 1.5,
        weight_unit: "kg",
        specs: {
          color: "Green",
          size: "Small",
          weight: "1.5kg",
          available_regions: ["US", "CA", "JP"],
        },
      },
    ],
  },
};
