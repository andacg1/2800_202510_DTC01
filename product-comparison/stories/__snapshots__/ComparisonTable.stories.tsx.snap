// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Comparison/ComparisonTable Empty smoke-test 1`] = `
<div data-theme="light"
     style="padding: 1rem;"
>
  <div>
    <div class="product-selection">
      <h3>
        Select Products to Compare (max 2)
      </h3>
      <div class="products-grid">
      </div>
    </div>
  </div>
</div>
`;

exports[`Product Comparison/ComparisonTable WithThreeProducts smoke-test 1`] = `
<div data-theme="light"
     style="padding: 1rem;"
>
  <div>
    <div class="product-selection">
      <h3>
        Select Products to Compare (max 2)
      </h3>
      <div class="products-grid">
        <div class="product-card ">
          <h4>
            Product A
          </h4>
          <div class="select-indicator">
            +
          </div>
        </div>
        <div class="product-card ">
          <h4>
            Product B
          </h4>
          <div class="select-indicator">
            +
          </div>
        </div>
        <div class="product-card ">
          <h4>
            Product C
          </h4>
          <div class="select-indicator">
            +
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Product Comparison/ComparisonTable WithTwoProducts smoke-test 1`] = `
<div data-theme="light"
     style="padding: 1rem;"
>
  <div>
    <div class="product-selection">
      <h3>
        Select Products to Compare (max 2)
      </h3>
      <div class="products-grid">
        <div class="product-card ">
          <h4>
            Product A
          </h4>
          <div class="select-indicator">
            +
          </div>
        </div>
        <div class="product-card ">
          <h4>
            Product B
          </h4>
          <div class="select-indicator">
            +
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
